Handle new build webhooks that use secretRefs instead of inline secrets

jhadvig opened the pull request:
Adding functionality to handle the new build webhooks that use secretRefs instead of inline secrets.
Now when user will want to add webhook to build config he will have to choose from available secrets that contain the `WebHookSecretKey` key.
Screen:
![a032d8ea-df07-4192-yd08-fb5c7e9ffd9f](https://user-images.githubusercontent.com/1668218/34688709-4ccff7f0-f4b3-11e7-9c65-3fe7aba96a7a.gif)


To be consistent with changes in CLI (introduced in https://github.com/openshift/origin/pull/17314) I've updated the `webhookURL` so it will show webhook urls as:
`https://localhost:9000/namespaces/myproject/buildconfigs/test-build/webhooks/<secret>/github`

Wasn't really sure if part of this [card](https://trello.com/c/lo6HwQgu/1106-2-build-webhooks-with-secretrefs) is also update the create-secret page to be able to create secrets that user could add to the webhook in the BC editor. 

@spadgett PTAL

jhadvig commented:
cc'ing @jwforres

jwforres commented:
I think its reasonable to allow them to pick an existing secret, but I think we definitely need a really simple "generate one for me" option.

jwforres commented:
Make sure you still handle if they have existing webhooks using the old embedded secret format. Don't let them add them that way, but don't break if they exist. They should be able to see and remove those webhooks.

Also this affects more than just the BC editor. You need to worry about the app generation code when you create an app from a builder image, that creates webhooks in the advanced settings (the code exists in two places, in both the catalog and in the console).  It might be better to handle app generation as a separate PR though.

jhadvig commented:
@jwforres I've update the PR.
In the create secret page I've added option to create webhook secret reference, that user can afterwards feed to the webhook trigger in the BS editor.
![secref2](https://user-images.githubusercontent.com/1668218/34728854-5289e5a4-f55b-11e7-94a8-2a67786c85c0.gif)


Also in the BC editor I've extracted the adding webhook trigger logic into a separate directive `oscWebhookTriggers` (similar to `oscSecrets` && `oscSourceSecrets`), so the user can create new webhook secret reference directly in editor if needed.
![secref1](https://user-images.githubusercontent.com/1668218/34728861-574cceee-f55b-11e7-9cfe-be967a3220ec.gif)


Already working on the app generation code as you mentioned in https://github.com/openshift/origin-web-console/pull/2640#issuecomment-356117314. Shall I open a separate PR or put it here ?

jwforres commented:
for creating a webhook secret key, it would be nice if there was a "generate" button next to the field similar to what some password managers do, that generates a value for you simillar to what we used to generate for webhooks before.  Since a lot of folks just want a random token in there.

jhadvig commented:
@jwforres I've update the PR
Added the 'Generate' btn:
![generate](https://user-images.githubusercontent.com/1668218/34774990-d79e7d66-f611-11e7-86e4-dc513ee025ef.png)

Also added logic to warn user that the URL is incomplete, since user doesnt have permissions to list Secrets. For that I've wrapped the `copy-to-clipboard` directive that we ued with the `webhookURL` filter into new `copyWebhookUrl` that contains the warning, to make it more clean and dont repeat the code.
In case the webhook trigger contains the old plain secret the webhook will be assembled as before.
![cantshowwebhook](https://user-images.githubusercontent.com/1668218/34775402-5a003d20-f613-11e7-8661-9154e6d5d80f.png)

jwforres commented:
So I see the alert where you tried to fetch the secrets, you should only
try to fetch the secrets if the canI list secrets check passes.

On Wed, Jan 10, 2018 at 8:35 AM, Jakub Hadvig <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> I've update the PR
> Added the 'Generate' btn:
> [image: generate]
> <https://user-images.githubusercontent.com/1668218/34774990-d79e7d66-f611-11e7-86e4-dc513ee025ef.png>
>
> Also added logic to warn user that the URL is incomplete, since user
> doesnt have permissions to list Secrets. For that I've wrapped the
> copy-to-clipboard directive that we ued with the webhookURL filter into
> new copyWebhookUrl that contains the warning, to make it more clean and
> dont repeat the code.
> In case the webhook trigger contains the old plain secret the webhook will
> be assembled as before.
> [image: cantshowwebhook]
> <https://user-images.githubusercontent.com/1668218/34775402-5a003d20-f613-11e7-8661-9154e6d5d80f.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/2640#issuecomment-356603719>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7RdX0gVqx3UXx_54b2exzo5EE5u3ks5tJLypgaJpZM4RW2J_>
> .
>

jhadvig commented:
@jwforres added the `canI` check before fetching the secrets

spadgett commented:
Any reason not to just generate the secret name and webhook key instead of making the user pick?

jwforres commented:
@spadgett you might actually want to re-use a webhook secret across different build configs

ncameronbritt commented:
Per our discussion on IRC:
![screen shot 2018-01-10 at 12 03 21 pm](https://user-images.githubusercontent.com/8974593/34785591-99b090b4-f5ff-11e7-8db4-0f8c8233b930.png)

jhadvig commented:
@jwforres I've update the PR per @ncameronbritt comment with the visual.
Here is a simple demo:
![secref3](https://user-images.githubusercontent.com/1668218/34827417-60c4a49e-f6da-11e7-8c83-1a4a60643cc2.gif)


PTAL

ncameronbritt commented:
Looks good @jhadvig!

spadgett commented on app/scripts/directives/oscWebhookTriggers.js:
```diff
@@ -0,0 +1,153 @@
+        };
+
+        // Check if new or modified webhook trigger is a duplicate. If so, show a warning under the appropriate trigger
+        var duplicityCheck = function(trigger) {
```
checkIfDuplicate

spadgett commented on app/scripts/directives/oscWebhookTriggers.js:
```diff
@@ -0,0 +1,153 @@
+
+          // After removing trigger check check other triggers for duplicates.
+          _.each($scope.webhookTriggers, function(trigger) {
+            if (trigger.isDuplicit) {
```
You always need to check duplicates since the `isDuplicate` will be false the first time it appears.

```
trigger A
trigger B (duplicate of A)
```

We need to unflag trigger B if trigger A is removed.

jhadvig replied:
Thats what Im doing in the `checkDuplicates` (renamed `duplicityCheck`) function which I feed with trigger I want to check, the one that is removed or the one user modifies 
Basically what Im doing there filtering an array of `matchingTriggers`  where all are basically duplicates, and I set the first ones `isDuplicate` field to `false` and on all others I set the field to `true`

spadgett replied:
I see. I missed that you are looping over all of them and not just checking the is duplicate flag on the one you're removing in this spot.

jhadvig replied:
yes, already updated the logic and will only loop on the trigger the action is on

spadgett commented on app/scripts/directives/oscWebhookTriggers.js:
```diff
@@ -0,0 +1,153 @@
+        // Check if new or modified webhook trigger is a duplicate. If so, show a warning under the appropriate trigger
+        var duplicityCheck = function(trigger) {
+          var matchingTriggers = _.filter($scope.webhookTriggers, function(wehbookTrigger) {
+            return _.isEqual(wehbookTrigger.data, trigger.data);
```
Does `data` include the secret? Do we expect to have duplicates often?

I'm wondering if we really need to be checking duplicates. @jwforres opinion?

jhadvig replied:
yes it does.
After speaking with @jwforres and @ncameronbritt we decided to notify user about the dups.

spadgett commented on app/scripts/directives/oscWebhookTriggers.js:
```diff
@@ -0,0 +1,153 @@
+          });
+
+          if (matchingTriggers.length > 1) {
+            _.first(matchingTriggers).isDuplicit = false;
```
```js
// Mark all except the first as duplicates.
_.each(matchingTriggers, function(trigger, i) {
  var first = i === 0;
  trigger.isDuplicate = !first;
});
```

spadgett commented on app/scripts/directives/oscWebhookTriggers.js:
```diff
@@ -0,0 +1,153 @@
+              return newTriggerType;
+            }
+            return key;
+          });
```
I think we're getting fancier than we need to.

```js
trigger.data[newTriggerType] = trigger.data[lastTriggerType];
delete trigger.data[lastTriggerType];
trigger.lastTriggerType = trigger.data.type;
```

spadgett commented on app/scripts/directives/oscWebhookTriggers.js:
```diff
@@ -0,0 +1,153 @@
+        $scope.checkLastAndAddNew = function() {
+          var lastTrigger = _.last($scope.webhookTriggers);
+          var lastTriggerSecretData = $filter('getWebhookSecretData')(lastTrigger);
+          if (lastTrigger.data.type && (lastTriggerSecretData.secret || lastTriggerSecretData.secretReference.name)) {
```
Isn't this a runtime error if secretReference isn't defined?

jhadvig replied:
The case shouldnt happen, since `lastTriggerSecretData.secret` or `lastTriggerSecretData.secretReference.name` should be set. But changing the check to `_.has()`

spadgett commented on app/scripts/directives/oscWebhookTriggers.js:
```diff
@@ -0,0 +1,153 @@
+          duplicityCheck(trigger);
+        };
+
+        var addWebhookTrigger = function() {
```
I'd probably name this `addEmptyWebhookTrigger`

spadgett commented on app/scripts/directives/util.js:
```diff
@@ -57,6 +57,28 @@ angular.module('openshiftConsole')
       }
     };
   })
+  .directive('copyWebhookUrl', function() {
```
Should be a component.

spadgett commented on app/scripts/filters/resources.js:
```diff
@@ -172,16 +172,30 @@ angular.module('openshiftConsole')
+          // arbitrarily many subresources can be included
+          // url encoding of the segments is handled by the url() function
+          // subresource segments cannot contain '/'
+          resource: "buildconfigs/webhooks/" + secret + "/" + type.toLowerCase(),
```
encodeURIComponent(secret)

spadgett replied:
Let's also encode the type to be safe

spadgett commented on app/scripts/filters/util.js:
```diff
@@ -322,6 +322,11 @@ angular.module('openshiftConsole')
   })
+  .filter("getWebhookSecretData", function() {
+    return function(webhookTrigger) {
+      return webhookTrigger.data[_.toLower(webhookTrigger.data.type)];
```
Probably better to use _.get to avoid errors if webhookTrigger is not defined

```js
var type = _.get(webhookTrigger, 'data.type');
if (!type) {
  return null;
}
return _.get(webhookTrigger, ['data', _.toLower(type)]);
```

spadgett commented on app/scripts/services/secrets.js:
```diff
@@ -141,8 +145,22 @@ angular.module("openshiftConsole")
 
+    var getWebhookSecretValue = function(secret, webhookSecrets) {
+      // 'secretReference' field is an indicator the the webhook is using SecretRef
+      // otherwise the secret is in the depricated plain form.
```
typo: deprecated

spadgett commented on app/views/browse/build-config.html:
```diff
@@ -386,28 +386,28 @@ <h3>Triggers <a href="{{'build-triggers' | helpLink}}" target="_blank"><span cla
                               </dt>
-                              <dd>
-                                <copy-to-clipboard clipboard-text="buildConfig.metadata.name | webhookURL : trigger.type : trigger.github.secret : project.metadata.name"></copy-to-clipboard>
+                              <dd class="with-warning">
```
I don't see where `with-warning` styles are added. Why only for GitHub webhooks?

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -47,268 +47,296 @@
-            id="username"
-            name="username"
-            ng-model="newSecret.data.username"
+            id="webhookSecretKey"
```
nit: we usually kebab case ids (webhook-secret-key)

rhamilto replied:
Need to check all id/for attributes as most of them are camel and not kebab.

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -47,268 +47,296 @@
-            <input type="checkbox" ng-model="add.cacert">
-            Use a custom ca.crt file
-          </label>
+            aria-describedby="username-help"
```
username-help ?

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -47,268 +47,296 @@
-          Private SSH key file for Git authentication.
+        <div class="help-block" id="secret-name-help">
+          Value of the secret will be supplied when invoking the webhook.
+          <a href="{{'webhooks' | helpLink}}" target="_blank"><span class="learn-more-inline">Learn More&nbsp;<i class="fa fa-external-link" aria-hidden="true"></i></span></a>
```
nit: ng-href

spadgett commented on app/views/directives/next-steps.html:
```diff
@@ -89,7 +89,7 @@ <h2 class="h3">Making code changes</h2>
       </span>
     </p>
-    <copy-to-clipboard clipboard-text="$ctrl.createdBuildConfig.metadata.name | webhookURL : trigger.type : trigger.github.secret : $ctrl.projectName"></copy-to-clipboard>
+    <copy-to-clipboard clipboard-text="$ctrl.createdBuildConfig.metadata.name | webhookURL : trigger.type :$ctrl.projectName : trigger.github"></copy-to-clipboard>
```
Should we be passing secrets here?

jhadvig replied:
we are, I've just have swapped the argument place, cause was adding additional arguman, but that wasnt necessary, will turn it back

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+        ng-model="trigger.data.type"
+        ng-disabled="isDeprecated(trigger)"
+        on-select="triggerTypeChange(trigger)"
+        on-remove="test()"
```
test?

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+        ng-disabled="!trigger.data.type"
+        ng-required="trigger.data.type"
+        on-select="triggerSecretChange(trigger)"
+        title="Select a webhooke secret reference"
```
typo: webhook

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+        on-select="triggerTypeChange(trigger)"
+        on-remove="test()"
+        search-enabled="false"
+        title="Select a webhooke type"
```
typo: webhook

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+        <ui-select-match placeholder="Webhook secret reference">
+          {{ $select.selected.metadata.name }}
+        </ui-select-match>
+        <ui-select-choices repeat="webhookSecret.metadata.name as webhookSecret in (webhookSecrets | filter : $select.search)">
```
track by (webhookSecret | uid)

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+              title="Toggle Password Visibility"
+              aria-label="Toggle Password Visibility"
+              ng-click="togglePassword($index)">
+            <i class="glyphicon glyphicon-eye-open"></i>
```
aria-hidden="true"

We probably shouldn't use `glyphicon`

rhamilto replied:
I think glyphicons are fine since they ship with Bootstrap/PatternFly.

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+          select-on-focus
+          disabled>
+        <span class="input-group-btn">
+          <a
```
button instead of a?

jhadvig replied:
button would submit the form in which we put the directive/component
we are using the anchor tag in the [copy-to-clipboard](https://github.com/openshift/origin-web-console/blob/master/app/views/directives/_copy-to-clipboard.html#L14) directive as well

spadgett replied:
We're trying to move to buttons when it makes sense because they're better for accessibility.

Does adding `type="button"` fix the form submit problem?

jhadvig replied:
yes, updated !

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+      </div>
+
+      <div class="remove-webhook-trigger-action">
+        <a ng-click="removeWebhookTrigger($index)"  href="" role="button" class="btn-remove">
```
button instead of a?

jhadvig replied:
Was inpired by what we are using in envVar editor and on other places like https://github.com/openshift/origin-web-console/blob/master/app/views/directives/osc-source-secrets.html#L100-L102

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+  </div>
+
+  <div class="add-webhook-action-btns">
+    <a href="" role="button" ng-click="checkLastAndAddNew()">Add Webhook</a>
```
```html
<button class="btn btn-link" ng-click="checkLastAndAddNew()">Add Webhook</button>
```

spadgett commented on app/views/directives/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,82 @@
+    <a href="" role="button" ng-click="checkLastAndAddNew()">Add Webhook</a>
+    <span ng-if="secretsVersion | canI : 'create'">
+      <span class="action-divider" aria-hidden="true"> | </span>
+      <a href="" role="button" ng-click="openCreateWebhookSecretModal()">Create New Webhook Secret</a>
```
button class="btn btn-link"

spadgett commented:
@rhamilto do you mind looking at the CSS / markup?

rhamilto commented on app/views/directives/create-secret.html:
```diff
@@ -47,268 +47,296 @@
-            Use a custom .gitconfig file
-          </label>
-        </div>
+        <label for="authentificationType">Authentication Type</label>
```
I see you didn't create this, but can you fix it?  Change to kebab case and add an input-id to the ui-select with the same kebab cased value.  It won't actually work because of https://github.com/angular-ui/ui-select/issues/1912, but at least the code will be right once the bug is resolved.

rhamilto commented on app/views/directives/create-secret.html:
```diff
@@ -47,268 +47,296 @@
+      <div ng-if="newSecret.authType === 'kubernetes.io/basic-auth'">
+        <div class="form-group">
+          <label for="username">Username</label>
+          <div>
```
Extraneous div?

rhamilto commented on app/views/directives/create-secret.html:
```diff
@@ -47,268 +47,296 @@
-        <div class="help-block">
-          Username is required.
+        <div class="form-group" ng-if="add.cacert" id="cacert">
+          <label class="required" for="cacert">CA Certificate File</label>
```
for attribute value does not match id value of input

jhadvig replied:
It is... under the `label` tag there is a `osc-file-input` directive to which we pass appropriate id in `drop-zone-id`

rhamilto commented on app/views/directives/create-secret.html:
```diff
@@ -47,268 +47,296 @@
-            required>
+      <div ng-if="newSecret.authType === 'kubernetes.io/ssh-auth'">
+        <div class="form-group" id="private-key">
+          <label for="privateKey" class="required">SSH Private Key</label>
```
for attribute value does not match id value of input

jhadvig replied:
It is... under the `label` tag there is a `osc-file-input` directive to which we pass appropriate id in `drop-zone-id`

rhamilto commented on app/styles/_forms.less:
```diff
@@ -7,6 +7,15 @@
+  .webhook-secrets-warning {
+    cursor: help;
+    padding: 5px;
+    font-size: 15px;
```
Nit: alphabetization

rhamilto commented on app/styles/_core.less:
```diff
@@ -1044,30 +1044,32 @@ a.disabled-link {
+        padding-right: 5px;
+        width: 50%;
+      }
+      .remove-webhook-trigger-action {
```
Nit: alpha

jhadvig commented:
@spadgett @rhamilto thanks for comments. I've addressed them.
Rewrite `oscWebhooTriggers` and `copyWebhookUrl` into components.
Regarding `stringData` usage, I've just changed the creation of the webhook secters. For other secret types I will open a card and assign it to myself, but didnt wanted to make this PR any more complex.

jwforres commented on app/components/osc-webhook-triggers/osc-webhook-triggers.html:
```diff
@@ -0,0 +1,81 @@
+          <button
+            type="button"
+            class="btn btn-default toggle"
+            title="Toggle Password Visibility"
```
its a nit, but this isnt really a password, i would probably call it a token

jhadvig commented:
@jwforres comment addressed. PTAL :)

spadgett commented:
/lgtm

spadgett commented:
/lgtm

spadgett commented:
/lgtm

spadgett commented on app/components/copy-webhook-url/copy-webhook-url.component.js:
```diff
@@ -0,0 +1,20 @@
+      triggerType: "=",
+      projectName: "=",
+      secret: "=",
+      webhookSecrets: "="
```
For future reference, these should probably be one way bindings with `<` instead of `=`. But I don't want to hold the PR on that.
